### ISSUE TITLE
Towards unified image builders

### ISSUE THREAD

**Issue by msanft** [NixOS/nixpkgs#324817]
## tl;dr

NixOS has many image builders that build an image file from a NixOS configuration, such as `image.repart`, `system.build.isoImage`, `system.build.sdImage`. Not only do these inconsistent interfaces create a bad UX (options being named differently but doing the same, etc.), they also make it hard to introduce new features for all of those builders due to duplication. With this, I propose to shift towards a unified image builder using `systemd-repart`. (i.e. the existing `image.repart` builder, with abstractions and default options that enable convenient builds of regularly built image types, such as the existing `system.build.isoImage` and `system.build.sdImage`). In addition to this proposition I'd like this issue to be a place for discussion on how such an interface should look like, ideally leading to an RFC.

## The Status Quo
We currently have (at least) 3 different publicly exposed builders that will emit a NixOS image based on a NixOS configuration:
- [`image.repart`](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/image/repart.nix): Arguably, this is the state-of-the-art builder, as it's the most configurable one and allows for the easiest attribute-based customizations. It can build almost any sort of image and is especially well-suited for immutable appliance images. It currently lacks support for image pre- and post-processing, for example to insert a ISO9660 (El Torito) boot record for making CD-ROM images. Furthemore, it's the only builder supporting discoverable disk images (DDIs), which are the de-facto standard of systemd nowadays.
- [`system.build.isoImage`](https://github.com/NixOS/nixpkgs/blob/4284c2b73c8bce4b46a6adf23e16d9e2ec8da4bb/nixos/modules/installer/cd-dvd/iso-image.nix#L3): The most popular builder for "desktop" installers and users that want an image that "just boots" and resembles closely what they are used to from other distributions. It did earn quite a lot of configurability over the years and has quite many options now. Uses an [imperative script](https://github.com/NixOS/nixpkgs/blob/master/nixos/lib/make-iso9660-image.sh) to build the actual image and lacks support for declarative partitioning, which is to be done after first boot in this scenario.
- [`system.build.sdImage`](https://github.com/NixOS/nixpkgs/blob/4284c2b73c8bce4b46a6adf23e16d9e2ec8da4bb/nixos/modules/installer/sd-card/sd-image.nix#L173): Quite similar to the ISO image builder, this one, again, has a new build script. It's the most popular choice for building tiny images, e.g. for embedded devices, as it has support for minimizing the partition sizes at first boot and writes a special `/boot/firmware` directory with custom bootloader contents. It lacks configurability for the actual partitions though, as root is always ext4, and `/boot/firmware` is always VFAT.

 Other than that, there also are private builders such as [`lib/make-disk-image.nix`](https://github.com/NixOS/nixpkgs/blob/master/nixos/lib/make-disk-image.nix), which are used throughout nixpkgs, e.g. to build VM images. Most of them are very use-case specific, but also encode a lot of logic.
 
## Why is this an Issue?
Maintaining multiple builders doing similar things is tedious, especially if their work can also be done by other builders. Furthermore, all builders have wildly different configuration interfaces, making it hard for users to find which knobs they need to turn to get what they want, and even finding the correct builder for the use-case in the first place can be a hassle. Furthermore, there are features which would be great to implement for all NixOS image builders, such as build-time execution of activation scripts, making image-based appliances more immutable by not requiring a writable root filesystem.

## What is desirable?
Part of this should be discussed within this issue. Personally, I propose a unified interface for building images (e.g. `build.image`) that should cover all current use-cases. The builder should be based on the current `image.repart`, as it provides the most configurability, and declarative partitioning. It also blends in very well with other systemd components, such as [`systemd-cryptsetup`](https://www.freedesktop.org/software/systemd/man/latest/systemd-cryptsetup.html). There should still be convenience targets (think a `system.build.isoImage`, but based on the unified builder interface), which should be compatible with the current builder interface. For continuous development of such a unified interface, a first step could port a basic builder, and then the exisiting builders (e.g. `system.build.isoImage`) could be ported from their existing build scripts to the unified builder, maintaining compatibility. This would be a huge step for both installation- as well as image-based NixOS, increasing nixpkgs-internal development velocity on these topics, many of which are very desirable for NixOS in enterprise use-cases.

**Comment by eclairevoyant** [comment]
Unifying 3+ interfaces that do the same thing sounds pretty uncontroversial, doubt it needs an rfc :) Thanks for raising this.

**Comment by msanft** [comment]
> Unifying 3+ interfaces that do the same thing sounds pretty uncontroversial, doubt it needs an rfc :) Thanks for raising this.

Yes, I think this part is out of question. 
The "RFC" part I was referring to rather meant a "discussion with an outcome" about how that *new* interface should look like

**Comment by phaer** [comment]
> Other than that, there also are private builders such as [lib/make-disk-image.nix](https://github.com/NixOS/nixpkgs/blob/master/nixos/lib/make-disk-image.nix), which are used throughout nixpkgs, e.g. to build VM images. Most of them are very use-case specific, but also encode a lot of logic.

I think more unification between disk-builders is a good idea as well. One thing that comes to mind, is that unless I am mistaken systemd-repart doesn't really support ZFS and such (yet)?
Ideally a generic solution would support that to be usable in vm tests and such, but for the same reason focusing on system.build.{isoImage,sdcard} first sounds good IMO.

(I added the significant label to give this more visibility. Please remove if its not appropriate here. I believe there's at least one older issue on this topic but wasn't able to find it)

**Comment by msanft** [comment]
> is that unless I am mistaken systemd-repart doesn't really support ZFS and such (yet)?

Correct. I should've probably mentioned that in the initial post, but didn't think of it. repart does, however, support btrfs, which should be able to handle _most_ ZFS use-cases too. It would also not be impossible to add a custom "post-processor" that populates a blank repart partition with a ZFS. I think it should be quite easy actually.

My implementation plan is to _create_ the general new interface (i.e. `build.image`) first, then migrate existing builders one-by-one, with enough of a waiting period in between to catch potential hiccups. 

Thanks for adding the label!

**Comment by arianvp** [comment]
`make-disk-image.nix` seems to be in a little bit of rough shape. I'm running into flakiness due to its use of `cptofs` https://github.com/NixOS/nixpkgs/issues/345492

Given all uses of `make-disk-image.nix` inside of nixpkgs all use `ext4` I'm curious if we can unify it with other stuff.

Idealy I'd switch to `repart` but I'd need some features added on top of repart:

Support grub in hybrid  mode (needed for cloud images) as they don't support UEFI boot often.
In general; unified support for "installing" a bootloader.

I wonder if the quickest way for this would be to write a tiny `runInLinuxVM` wrapper that calls `NIXOS_INSTALL_BOOTLOADER=1 switch-to-configuration boot`.


It's not ideal (we can probably install bootloaders without needing a VM) but we know it works.

**Comment by arianvp** [comment]
> No MBR/hybrid MBR support means we can't use repart generated images on older SBCs that do not support booting from GPT, like the Raspberry Pi 02w and 3B+, without [additional post-processing](https://github.com/Electrostasy/dots/blob/978d1fb0e8e64bc10ac1e21f154682837a66e4b0/hosts/deimos/image.nix#L84-L91). As far as I can tell it's on the [TODO](https://github.com/systemd/systemd/blob/26e6986527fe553edf35a442a520deb10c45a155/TODO#L2421-L2426).

Hybrid support shouldn't be too hard I think? It's a matter of creating a grub partition 

https://github.com/systemd/mkosi/blob/main/mkosi/__init__.py#L2993

And then we need to install grub in the final image (Could e.g. be done using a `runInLinuxVM`)

**Comment by Electrostasy** [comment]
> Hybrid support shouldn't be too hard I think? It's a matter of creating a grub partition
> 
> https://github.com/systemd/mkosi/blob/main/mkosi/__init__.py#L2993
> 
> And then we need to install grub in the MBR in the final image (Could e.g. be done using a `runInLinuxVM`)

Interesting, I may need to experiment with this. But isn't the resulting image still GPT? For the purposes of getting SBCs that can't boot from GPT to boot from GPT, just creating a hybrid MBR is easy enough with a configurable post-process stage IMO. From my understanding, if you mess with the partition table afterwards (resizing/adding partitions?), it can break in unexpected ways if not done correctly (so it can be brittle). Ideally we wouldn't need hybrid MBRs in the first place.

> Yes this was the primary purpose it was developed for in the first place! If you pass it an image with `--image` that already has a GPT; it will just do purely additive operations. We could perhaps add an option to add a base image to the module?

This would be awesome to have and would probably solve my gripes with hardcoded u-boot offsets mentioned earlier.

**Comment by arianvp** [comment]
> Interesting, I may need to experiment with this. But isn't the resulting image still GPT?

It's specific to grub I think.   You put (part of) GRUB in the MBR;   that code the loads the rest of the GRUB bootloader from the GRUB partition.  And then the grub bootloader reads the kernels and initrds from the ESP

https://en.wikipedia.org/wiki/BIOS_boot_partition

**Comment by DaanDeMeyer** [comment]
@msanft If you're actually able to figure out El Torito, we wouldn't mind supporting that in repart itself (directly, no `xorisso` or similar tools allowed)
